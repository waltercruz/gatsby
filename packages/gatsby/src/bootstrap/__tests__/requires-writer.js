const { joinPath } = require(`gatsby-core-utils`)
const requiresWriter = require(`../requires-writer`)

const now = Date.now()

const generatePagesState = pages => {
  let state = new Map()
  pages.forEach(page => {
    state.set(page.path, {
      component: ``,
      componentChunkName: ``,
      ...page,
    })
  })

  return state
}

jest.mock(`fs-extra`, () => {
  return {
    writeFile: () => Promise.resolve(),
    move: () => {},
  }
})

const mockFsExtra = require(`fs-extra`)

describe(`requires-writer`, () => {
  const program = {
    directory: `/dir`,
  }
  let originalDateNow = global.Date.now

  beforeEach(() => {
    global.Date.now = () => now
    requiresWriter.resetLastHash()
  })

  afterAll(() => {
    global.Date.now = originalDateNow
  })

  describe(`writeAll`, () => {
    it(`writes requires files`, async () => {
      const pages = generatePagesState([
        {
          component: `component1`,
          componentChunkName: `chunkName1`,
          matchPath: `matchPath1`,
          path: `/path1`,
        },
        {
          component: `component2`,
          componentChunkName: `chunkName2`,
          path: `/path2`,
        },
      ])

      const spy = jest.spyOn(mockFsExtra, `writeFile`)
      await requiresWriter.writeAll({
        pages,
        program,
      })

      expect(spy).toBeCalledWith(
        joinPath(`/dir`, `.cache`, `match-paths.json.${now}`),
        JSON.stringify([{ path: `/path1`, matchPath: `matchPath1` }], null, 4)
      )
    })
  })

  describe(`matchPath`, () => {
    let matchPaths = []

    beforeEach(() => {
      mockFsExtra.writeFile.mockImplementation((file, buffer) => {
        if (file.includes(`match-paths.json`)) {
          matchPaths = JSON.parse(String(buffer))
        }

        return Promise.resolve()
      })
    })

    it(`should sort matchPaths by specificity`, async () => {
      const pages = generatePagesState([
        {
          path: `/`,
        },
        {
          path: `/app/`,
          matchPath: `/app/*`,
        },
        {
          path: `/app/projects/`,
          matchPath: `/app/projects/*`,
        },
        {
          path: `/app/clients/`,
          matchPath: `/app/clients/*`,
        },
        {
          path: `/app/login/`,
        },
      ])

      await requiresWriter.writeAll({
        pages,
        program,
      })

      expect(matchPaths[0].path).toBe(pages.get(`/app/clients/`).path)
      expect(matchPaths[matchPaths.length - 1].path).toBe(
        pages.get(`/app/`).path
      )
      expect(matchPaths).toMatchSnapshot()
    })
  })
})
