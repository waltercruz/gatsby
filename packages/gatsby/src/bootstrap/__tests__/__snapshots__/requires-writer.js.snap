// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requires-writer matchPath should sort matchPaths by specificity 1`] = `
Array [
  Object {
    "matchPath": "/app/clients/*",
    "path": "/app/clients/",
  },
  Object {
    "matchPath": "/app/projects/*",
    "path": "/app/projects/",
  },
  Object {
    "matchPath": "/app/*",
    "path": "/app/",
  },
]
`;
